### PR TITLE
Add configurable gyro aiming and tilt steering

### PR DESCRIPTION
**Pull request by Nightwalker743**
## Description

Adds configurable device gyro input for aiming, camera control, movement, and steering.

This work is based on and substantially expands the original `gyro-aiming-v2` implementation by @unbelievableflavour.

## Recording

https://drive.google.com/file/d/1zCqSlLfZk-y7ZqP9Zux5Yn-2iXY92AW-/view?usp=drivesdk

## Type of Change

- [ ] Bug fix
- [ ] Performance / stability improvement
- [ ] Compatibility improvements
- [x] Other (requires prior approval)

## Checklist

- [x] If I have access to `#code-changes`, I have discussed this change there and it has been green-lighted. If I do not have access, I have still provided clear context in this PR. If I skip both, I accept that this change may face delays in review, may not be reviewed at all, or may be closed.
- [x] This change aligns with the current project scope (core functionality, stability, or performance). If not, it has been explicitly approved beforehand.
- [x] I have attached a recording of the change.
- [x] I have read and agree to the contribution guidelines in `CONTRIBUTING.md`.

---
## Summary by cubic
Adds configurable gyro aiming and tilt steering. Previously there was no gyro support; now device motion can drive mouse or stick output, with hold, toggle, or ratchet activation via a new `GYRO_MODIFIER` binding.

**New Features**
- Per-container settings for output mode, sensitivity, smoothing, axis inversion, and activation.
- Tilt steering sustains horizontal stick input at the device's held angle.
- Devices without a gyroscope show the feature as unavailable instead of erroring.

**Bug Fixes**
- Improves multi-controller input tracking and cleanup.
- Prevents stale touch actions and preserves control-profile updates more reliably.

<sup>Written for commit 9b15bc673e221d9c0786768d521a5ce0093aab5e. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/utkarshdalal/GameNative/pull/1882?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added gyroscope aiming for compatible devices, with mouse or stick output.
  * Added configurable activation modes, sensitivity, smoothing, tilt steering, inversion, and deadzones.
  * Added gyro controls and settings to the quick menu, including assignment and reset options.
  * Added support for devices without gyroscope hardware.
  * Added translations for gyro settings across multiple languages.

* **Bug Fixes**
  * Improved multi-controller input tracking and cleanup.
  * Prevented stale touch actions and preserved control-profile updates more reliably.

### CODE REVIEW

File: app/src/main/java/app/gamenative/ui/screen/xserver/PhysicalControllerHandler.kt [hunk 1]
```diff
@@ -707,16 +830,20 @@ class PhysicalControllerHandler(
                 else {
                     when (binding) {
                         Binding.GAMEPAD_LEFT_THUMB_UP, Binding.GAMEPAD_LEFT_THUMB_DOWN -> {
-                            state.thumbLY = if (isActionDown) offset else 0f
+                            state.thumbLY = gyroStickMixer?.invoke(binding, isActionDown, offset, sourceKeyCode)
```
[the remark this answers is not in the document]
Comment (Nightwalker743) [reply app/src/main/java/app/gamenative/ui/screen/xserver/PhysicalControllerHandler.kt]: This matches the pre-gyro shared gamepad behavior: simultaneous controllers writing the same axis are not aggregated per source. Fixing that would require a broader multi-controller input model and is outside this gyro regression fix.

File: app/src/main/java/app/gamenative/ui/screen/xserver/PhysicalControllerHandler.kt [hunk 2]
```diff
@@ -99,13 +143,31 @@ class PhysicalControllerHandler(
...
         clearMouseMoveContributions()
         clearScrollRepeats()
         activeSequenceTriggerBindings.clear()
         showKeyboardPressed = false
         closeRadialMenuIfOpen(commit = false)
+        sendGamepadState()
+    }
+
+    fun onInputDeviceRemoved(deviceId: Int) {
+        cancelActiveSequences()
```
[the remark this answers is not in the document]
Comment (Nightwalker743) [reply app/src/main/java/app/gamenative/ui/screen/xserver/PhysicalControllerHandler.kt]: This requires disconnecting one controller during another controller's short in-progress sequence. Device-scoped sequence ownership would be a broader multi-action change; the current cleanup deliberately cancels output to avoid stuck inputs, so this edge case is unchanged here.

File: app/src/test/java/app/gamenative/ui/screen/xserver/InputControlsProfileCopyTest.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.verify
+
+class InputControlsProfileCopyTest {
+    @Test
+    fun copyNotifiesViewThatProfileContentWasReplaced() {
+        val source = mock<ControlsProfile>()
+        val target = mock<ControlsProfile>()
+        val view = mock<InputControlsView>()
+        `when`(source.elements).thenReturn(arrayListOf<ControlElement>())
```
[the remark this answers is not in the document]
Comment (Nightwalker743) [reply app/src/test/java/app/gamenative/ui/screen/xserver/InputControlsProfileCopyTest.kt]: Addressed in 9b15bc673. The test now uses non-empty source and target profiles, verifies removal and addition, checks copied element fields, and verifies a distinct deep-copied radial menu.

File: app/src/main/res/values-de/strings.xml [hunk 4]
```diff
@@ -2363,4 +2363,43 @@
...
+    <string name="gyro_section_output">Ausgabe und Bewegung</string>
+    <string name="gyro_section_activation">Aktivierung</string>
+    <string name="gyro_section_response">Reaktion</string>
+    <string name="gyro_output">Ausgabe</string>
+    <string name="gyro_output_subtitle">Wähle aus, was die Gerätebewegung steuert.</string>
+    <string name="gyro_tilt_steering">Neigungslenkung</string>
+    <string name="gyro_tilt_steering_subtitle">Halte das Gerät geneigt, um eine horizontale Stick-Eingabe beizubehalten. Die aktuelle Position wird bei jeder Gyro-Aktivierung zentriert.</string>
+    <string name="gyro_tilt_unavailable">Dieses Gerät bietet nicht den für die Neigungslenkung erforderlichen Orientierungssensor.</string>
+    <string name="gyro_activation">Aktivierung</string>
+    <string name="gyro_activation_help">Halten, Umschalten und Ratsche verwenden eine Gyro zugewiesene Steuerung.</string>
```
[the remark this answers is not in the document]
Comment (Nightwalker743) [reply app/src/main/res/values-de/strings.xml]: Addressed in 9b15bc673.

File: app/src/main/java/app/gamenative/ui/component/QuickMenu.kt [hunk 5]
```diff
@@ -999,20 +1051,28 @@ fun QuickMenu(
...
-                                                    isActive = item.id in activeToggleIds,
+                                                    isActive = if (item.id == QuickMenuAction.GYRO) {
+                                                        gyroEnabled
+                                                    } else {
+                                                        item.id in activeToggleIds
+                                                    },
                                                     onClick = {
-                                                        if (onItemSelected(item.id)) {
+                                                        if (item.id == QuickMenuAction.GYRO && gyroMenu != null) {
+                                                            gyroMenu.setEnabled(!gyroEnabled)
```
[the remark this answers is not in the document]
Comment (Nightwalker743) [reply app/src/main/java/app/gamenative/ui/component/QuickMenu.kt]: This is intentional: keeping the menu open makes the settings gear available immediately after enabling gyro. Gyro is overlay-suppressed while the Quick Menu is visible, so tilt input does not move the game underneath it.

File: app/src/main/java/com/winlator/inputcontrols/ControlElement.java [hunk 6]
```diff
@@ -790,13 +825,17 @@ private static void setVerticalRangeOutlinePath(Path path, Rect boundingBox, flo
     public void draw(Canvas canvas) {
         int snappingSize = inputControlsView.getSnappingSize();
         Paint paint = inputControlsView.getPaint();
-        boolean active = selected || currentPointerId != -1;
+        boolean showGyroState = !inputControlsView.isEditMode()
+                && inputControlsView.isGyroEnabled() && isGyroModifierControl();
+        boolean gyroActive = showGyroState && inputControlsView.isGyroActive();
```
[the remark this answers is not in the document]
Comment (Nightwalker743) [reply app/src/main/java/com/winlator/inputcontrols/ControlElement.java]: This is intentional. Gyro activation is global, and every control assigned as a gyro modifier reflects that active state for hold, toggle, and ratchet modes.

File: app/src/test/java/com/winlator/widget/InputControlsViewStickMixingTest.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        assertEquals(0f, upReleased, 0f)
+    }
+
+    @Test
+    fun `releasing old direction does not clear newly active direction`() {
+        val rawDownSource = ExternalControllerBinding.getKeyCodeForAxis(MotionEvent.AXIS_Y, 1)
+        val rawUpSource = ExternalControllerBinding.getKeyCodeForAxis(MotionEvent.AXIS_Y, -1)
+
+        val down = InputControlsView.updatePhysicalBaseAxis(0f, true, 0.8f, rawDownSource)
+        val up = InputControlsView.updatePhysicalBaseAxis(down, true, -0.7f, rawUpSource)
```
[the remark this answers is not in the document]
Comment (Nightwalker743) [reply app/src/test/java/com/winlator/widget/InputControlsViewStickMixingTest.kt]: Addressed in 9b15bc673. The test now passes 0f directly and describes the replacement behavior accurately.

File: app/src/test/java/app/gamenative/data/GyroSettingsTest.kt [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+import org.junit.Assert.assertTrue
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.kotlin.any
+import org.mockito.kotlin.doAnswer
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.whenever
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
```
[the remark this answers is not in the document]
Comment (Nightwalker743) [reply app/src/test/java/app/gamenative/data/GyroSettingsTest.kt]: Retained Robolectric after verification: without it, Mockito cannot initialize/mock Container because of its Android-dependent class initialization. The plain JUnit version fails before the settings assertions run.

File: app/src/test/java/com/winlator/inputcontrols/RangeScrollerTest.kt [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            true
+        }
+        val element = mockk<ControlElement>(relaxed = true)
+        every { element.boundingBox } returns Rect(0, 0, 100, 100)
+        every { element.bindingCount } returns 1
+        every { element.range } returns ControlElement.Range.FROM_A_TO_Z
+        every { element.orientation } returns 0.toByte()
+        val scroller = RangeScroller(view, element)
+
+        scroller.handleTouchDown(50f, 50f)
```
[the remark this answers is not in the document]
Comment (Nightwalker743) [reply app/src/test/java/com/winlator/inputcontrols/RangeScrollerTest.kt]: Addressed in 9b15bc673. Long-press scheduling and elapsed time are now controlled by the test, with no Timer thread, wall-clock delay, or latch.
